src/System.Memory/src/System/Buffers/ReadOnlySequence.cs

```diff
@@ -286,7 +286,7 @@ public ReadOnlySequence<T> Slice(long start, SequencePosition end)
 
             int currentLength = startSegment.Memory.Length - (int)startIndex;
 
-            // Position in startSegment
+            // Position not in startSegment
             if (currentLength <= start)
```

Comment:
If I change this to `if (currentLength < start)`, the [new test fails](https://github.com/dotnet/corefx/pull/29205/files#diff-528f60ed131bab89f2a08f806c4b6d2bR105):
```text
System.Memory.Tests.ReadOnlySequenceTestsCommonByte.SliceStartAndEndPosition [FAIL]
        Assert.Equal() Failure
        Expected: BufferSegment`1 { Memory = System.ReadOnlyMemory<Byte>[10], Next = null, RunningIndex = 10 }
        Actual:   BufferSegment`1 { Memory = System.ReadOnlyMemory<Byte>[10], Next = BufferSegment`1 { Memory = System.ReadOnlyMemory<Byte>[10], Next = null, RunningIndex = 10 }, RunningIndex = 0 }
        Stack Trace:
           D:\GitHub\Fork\corefx\src\System.Memory\tests\ReadOnlyBuffer\ReadOnlySequenceTests.Common.byte.cs(97,0): at System.Memory.Tests.ReadOnlySequenceTestsCommonByte.SliceStartAndEndPosition()
```

Reply:
While Slice should return 2 segments? The second segment is empty. Return 1 segment is more optimal. Now you have the same bug but when end is first segment.

Reply:
> While Slice should return 2 segments? The second segment is empty. Return 1 segment is more optimal. 

It shouldn't return 2 segments and it doesn't. Right now it returns 1 segment. `sliced.Start` and `sliced.End` are both pointing to segment 2 if I leave things as is.

If I make the change you are suggesting, `slice.Start` would point to segment1 and `slice.End` would point to segment2.

Reply:
You wrote some set of strange tests. I can write more tests where current implementation have bug but i do not have time. Is it possible to get some grant to work on Microsoft project full time? I will program for food starting from $6 per hour.

Reply:
`buffer` = [x y z] -> [a b c]
`buffer.Slice(3)` = [a b c]
`buffer.Slice(3, buffer.End)` = [a b c]

`buffer.Slice(3)` should be equal to `buffer.Slice(3, buffer.End)`

If I change it to `if (currentLength < start)`, 
`buffer.Slice(3, buffer.End)` becomes =[] -> [a b c]

Reply:
Ok. You are right. Here should be `if (currentLength <= start)` but add special case when `sliceEndObject = startObject` or you have bug here.

Reply:
You call `SeekMultiSegment` with next segment that is after sliceEndObject

Reply:
If sliceEndObject == startObject, then `currentLength > start`. Otherwise, the `if (startRange + (ulong)start > sliceRange)` check will throw an exception.

Reply:
```
        [Fact]
        public static void Slice()
        {
            var segment1 = new BufferSegment<byte>(new byte[10]);
            BufferSegment<byte> segment2 = segment1.Append(new byte[10]);

            var buffer1 = new ReadOnlySequence<byte>(segment1, 0, segment1, 10);
            var buffer2 = new ReadOnlySequence<byte>(segment1, 0, segment2, 10);

            ReadOnlySequence<byte> sliced = buffer2.Slice(0, buffer1.End);
            Assert.Equal(0, sliced.Length);
}
```

Reply:
We don't support passing in a sequence position from one ReadOnlySequence into the other.
And, what would happen in the second case (`ReadOnlySequence<byte> sliced = buffer.Slice(0, buffer.Start);)` if we don't add these checks?

Reply:
Third test

```cs
        [Fact]
        public static void Slice()
        {
            var segment1 = new BufferSegment<byte>(new byte[10]);
            BufferSegment<byte> segment2 = segment1.Append(new byte[10]);

            var buffer = new ReadOnlySequence<byte>(segment1, 0, segment2, 10);

            // Next should throw exception but will not
            ReadOnlySequence<byte> sliced = buffer.Slice(10, buffer.Start);
       }
```

Reply:
```C#
[Fact]
public static void SliceMismatchSequencePositions()
{
    var segment1 = new BufferSegment<byte>(new byte[10]);
    BufferSegment<byte> segment2 = segment1.Append(new byte[10]);

    var buffer1 = new ReadOnlySequence<byte>(segment1, 0, segment1, 10);
    var buffer2 = new ReadOnlySequence<byte>(segment1, 0, segment2, 10);

    // mismatch of SequencePositions from different buffers is not supported
    ReadOnlySequence<byte> sliced = buffer2.Slice(10, buffer1.End);

    Assert.Equal(segment2, sliced.Start.GetObject());   // This mismatch will happen, what should happen?
    Assert.Equal(segment1, sliced.End.GetObject());     // This mismatch will happen, what should happen?

    // This is valid and passes as expected
    sliced = buffer1.Slice(10, buffer1.End);

    Assert.Equal(segment1, sliced.Start.GetObject());
    Assert.Equal(segment1, sliced.End.GetObject());

    // This is valid and passes as expected
    sliced = buffer2.Slice(10, buffer2.End);

    Assert.Equal(segment2, sliced.Start.GetObject());
    Assert.Equal(segment2, sliced.End.GetObject());
}

[Fact]
public static void SliceSequencePositionComparison()
{
    var segment1 = new BufferSegment<byte>(new byte[0]);
    BufferSegment<byte> segment2 = segment1.Append(new byte[10]);

    var buffer = new ReadOnlySequence<byte>(segment1, 0, segment2, 10);

    ReadOnlySequence<byte> sliced = buffer.Slice(0, buffer.Start);

    // This fails, sliced.Start.GetObject() == segment2 instead of segment1, currently
    // This is a known issue related to SequencePosition comparisons
    Assert.Equal(segment1, sliced.Start.GetObject());
    Assert.Equal(segment1, sliced.End.GetObject());

    Assert.Equal(0, sliced.Start.GetInteger());
    Assert.Equal(0, sliced.End.GetInteger());
}
```

Reply:
cc @KrzysztofCwalina, @GrabYourPitchforks